### PR TITLE
Various doc improvements

### PR DESCRIPTION
**Pull request by gsmet**
@geoand could you take a look at this one?

The first commit is an easy one.

As for the other commit, I entirely separated the application/extension cases so that we can point users writing applications to this documentation.

Prior to that, it was really targeting the extension developers.

Thanks!

### CODE REVIEW

**Comment by geoand** [comment]
Sure thing, I'll check it in a few minutes

File: docs/src/main/asciidoc/writing-native-applications-tips.adoc [hunk 1]
```diff
@@ -7,47 +7,85 @@ https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
 
 include::./attributes.adoc[]
 
-This guide contains various tips and tricks for getting around problems that might arise when attempting to run java applications as native executables.
+This guide contains various tips and tricks for getting around problems that might arise when attempting to run Java applications as native executables.
 
-== Register reflection
+Note that we differentiate two contexts where the solution applied might be different:
```
Comment (geoand) [inline docs/src/main/asciidoc/writing-native-applications-tips.adoc]: I love this change, makes things a lot more clear. :100:

File: docs/src/main/asciidoc/writing-native-applications-tips.adoc [hunk 2]
```diff
@@ -7,47 +7,85 @@ https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
...
-[source]
-----
-Caused by: org.xml.sax.SAXException: SAX2 driver class com.sun.org.apache.xerces.internal.parsers.SAXParser not found
-java.lang.ClassNotFoundException: com.sun.org.apache.xerces.internal.parsers.SAXParser
-at org.xml.sax.helpers.XMLReaderFactory.loadClass(XMLReaderFactory.java:230)
-----
+== Supporting native in your application
 
-If one were to execute the native executable generation process manually, the `-H:ReflectionConfigurationFiles=` flag would be used to point to a JSON configuration file that specifies the program elements that would be accessed reflectively.
+GraalVM comes with a number of constraints and making your application a native executable might require a few tweaks.
```
Comment (geoand) [inline docs/src/main/asciidoc/writing-native-applications-tips.adoc]: `GraalVM comes with a number` -> `GraalVM imposes a number` maybe?

`and making` -> `so making` perhaps?

File: docs/src/main/asciidoc/writing-native-applications-tips.adoc [hunk 3]
```diff
@@ -139,38 +219,115 @@ method com.amazonaws.services.s3.model.CryptoConfiguration.<init>(CryptoMode)
...
+While writing native application you'll need to define proxy classes at image build time by specifying the list of interfaces that they implement.
+
+In such a situation, the error you might encounter is:
+
+[source]
+----
+com.oracle.svm.core.jdk.UnsupportedFeatureError: Proxy class defined by interfaces [interface org.apache.http.conn.HttpClientConnectionManager, interface org.apache.http.pool.ConnPoolControl, interface com.amazonaws.http.conn.Wrapped] not found. Generating proxy classes at runtime is not supported. Proxy classes need to be defined at image build time by specifying the list of interfaces that they implement. To define proxy classes use -H:DynamicProxyConfigurationFiles=<comma-separated-config-files> and -H:DynamicProxyConfigurationResources=<comma-separated-config-resources> options.
+----
+
+Solving this issue requires to define add the `-H:DynamicProxyConfigurationResources=<comma-separated-config-resources>` option and to provide a dynamic proxy configuration file.
```
Comment (geoand) [inline docs/src/main/asciidoc/writing-native-applications-tips.adoc]: `to define add` -> `requires adding`

File: docs/src/main/asciidoc/writing-native-applications-tips.adoc [hunk 4]
```diff
@@ -139,38 +219,115 @@ method com.amazonaws.services.s3.model.CryptoConfiguration.<init>(CryptoMode)
...
+[NOTE]
+====
+More information about reflection in GraalVM can be found https://github.com/oracle/graal/blob/master/substratevm/REFLECTION.md[here].
+====
 
-For more information about `--delay-class-initialization-to-runtime`, please read https://medium.com/graalvm/understanding-class-initialization-in-graalvm-native-image-generation-d765b7e4d6ed[this blog post].
+=== Alternative with @RegisterForReflection
 
-== Proxy Classes management
+As for applications, you can also use the `@RegisterForReflection` annotation if the class is in your extension and not in a third-party jar.
```
Comment (geoand) [inline docs/src/main/asciidoc/writing-native-applications-tips.adoc]: `As for applications` -> `As is the case for applications`

File: docs/src/main/asciidoc/writing-native-applications-tips.adoc [hunk 5]
```diff
@@ -139,38 +219,115 @@ method com.amazonaws.services.s3.model.CryptoConfiguration.<init>(CryptoMode)
 Call path from entry point to com.amazonaws.services.s3.model.CryptoConfiguration.<init>(CryptoMode):
 ----
 
-If one were to execute the native executable generation process manually, the `--delay-class-initialization-to-runtime=com.amazonaws.services.s3.model.CryptoConfiguration` flag would need to be passed to `native-image` to configure runtime initialization for `CryptoConfiguration`.
+If you need to delay the initialization of a class, you can use the `--initialize-at-run-time=<package or class>` configuration knob.
 
-Quarkus simplifies things by allowing extensions authors to simply register a `RuntimeInitializedClassBuildItem`. A simple example of doing so could be:
+It should be added to the `native-image` configuration using an `<additionalBuildArg>` as for the examples above.
```
Comment (geoand) [inline docs/src/main/asciidoc/writing-native-applications-tips.adoc]: `as for the examples above` -> `as shown in the examples above`?

**Comment by geoand** [comment]
@gsmet I added some very minor comments. 
I think this is a great improvement!

File: docs/src/main/asciidoc/writing-native-applications-tips.adoc (hunk 2 above)
Comment (gsmet) [reply docs/src/main/asciidoc/writing-native-applications-tips.adoc]: applied the first one but not the second one. Don't know why but `and` sounds better to my ears.

File: docs/src/main/asciidoc/writing-native-applications-tips.adoc (hunk 3 above)
Comment (gsmet) [reply docs/src/main/asciidoc/writing-native-applications-tips.adoc]: Good catch.

File: docs/src/main/asciidoc/writing-native-applications-tips.adoc (hunk 5 above)
Comment (gsmet) [reply docs/src/main/asciidoc/writing-native-applications-tips.adoc]: Yes, better, applied.

File: docs/src/main/asciidoc/writing-native-applications-tips.adoc (hunk 2 above)
Comment (geoand) [reply docs/src/main/asciidoc/writing-native-applications-tips.adoc]: OK :)

**Comment by gsmet** [comment]
@geoand I force pushed an update, thanks for the suggestions.

**Review by geoand — APPROVED** [review]
:+1:

**Comment by geoand** [comment]
Merged :)

**Pull request merged**